src/js/server/lib/datasource.js

```diff
@@ -0,0 +1,161 @@
+// because it has a non-namespaced listener for `onRead`, which we cannot replace. The `gradeNames` and `listener` block
+// are copied from that grade.
+// 
+// To work with the value returned from CouchDB, you most likely will simply want to add another `onRead` listener with 
+// a lower priority, as in:
+//
+// listeners: {
+//   "onRead.logResponse": {
+//     funcName: "console.log",
+//     args: ["I received the following response:", "{arguments}.0"],
+//     priority: "after:gpii.express.user.couchdb.read"
+//   }
+// }
+//
+// For more information on priorities, see: http://docs.fluidproject.org/infusion/development/Priorities.html
+//
+"use strict";
+var fluid  = require("infusion");
+var gpii   = fluid.registerNamespace("gpii");
+
+require("kettle");
+
+fluid.registerNamespace("gpii.express.user.couchdb.read");
+
+// Replacement for a similar function in the default implementation in dataSource.js.  Adds support for rule-based 
+// decoding of couch response, which can work with lists, individual records, shows, whatever.
+gpii.express.user.couchdb.read.transformCouchResults = function (that, resp) {
+    // if undefined, pass that through as per dataSource (just for consistency in FS-backed tests)
+    var togo;
+    if (resp === undefined) {
+        togo = undefined;
+    } else {
+        if (resp.error) {
+            var error = {
+                isError: true,
+                message: resp.error + ": " + resp.reason
+            };
+            togo = fluid.promise();
+            togo.reject(error);
+        } else {
+            togo = fluid.model.transformWithRules(resp, that.options.rules.read);
+        }
+    }
+    return togo;
+};
+
+fluid.defaults("gpii.express.user.couchdb.read", {
+    gradeNames: ["kettle.dataSource.URL"],
+    writable:   false,
+    rules: {
+        read: {
+            "": ""
+        }
+    },
+    listeners: {
+        "onRead.transformCouchResults": {
+            funcName:  "gpii.express.user.couchdb.read.transformCouchResults",
+            args:      ["{that}", "{arguments}.0"],
+            namespace: "gpii.express.user.couchdb",
+            priority:  "after:encoding"
+        }
+    }
+});
+
+
+
+// A rewritten version of `kettle.dataSource.CouchDB.write` that decouples the read and write URLS (so that you can
+// read from a _view and write to a native Couch API.  The "read" URL is found at `options.urls.read`, and the "write" 
+// URL is set using `options.url.write`.   As this class  extends `kettle.dataSource.URL.writable`, `options.url` is 
+// set to the value of `options.url.write` by default.
+// 
+// Also allows rewriting of all payloads and responses using transformation rules.
+
+fluid.registerNamespace("gpii.express.user.couchdb.writable");
+
+// A replacement for the default couch "write" operation that removes the hard-coded `value` wrapper in favor of rules.
+gpii.express.user.couchdb.writable.transformModelToCouch = function (that, model, options) {
```

Comment:
I assume this is a good candidate for being replaced by your upcoming asymmetric dataSource